### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by resdenia**
Metrics from google services to logz.io
Using APIs:
- Cloud Function
- Monitoring API
- Job Scheduler

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,73 @@
...
+| listener_url | Use the listener URL specific to the region of your Logz.io account. You can look it up [here](https://docs.logz.io/user-guide/accounts/account-region.html). |
+| token        | The metrics' shipping token of the account you want to ship to.                                                                                               |
+| region       | Region where you want to upload Cloud Function. \*`Requires for Deploy to Cloud option for platform`.                                                         |
+
+| function_name | Function name will be using as Google Cloud Function name. (Default:`metrics_gcp`) |
+| metric_types | Will send metrics that match the Google metric type. Detailed list you can find [here](https://cloud.google.com/monitoring/api/metrics_gcp) (ex: `cloudfunctions.googleapis.com`) |
+
+## Check Logz.io for your metrics
+
+Give your metrics a few minutes to get from your system to ours, and then open [Kibana](https://app.logz.io/#/dashboard/metrics).
```
Comment (Doron-Bargo) [inline README.md]: Why open kibana?

File: run.sh [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+        is_error=true
+        echo -e "\033[0;31mrun.sh (1): Logz.io Token is missing please rerun the script with the relevant parameters\033[0;37m"
+    fi
+    if [[ -z "$region" ]]; then
+        is_error=true
+        echo -e "\033[0;31mrun.sh (1): Region for Google Cloud Platform is missing please rerun the script with the relevant parameters\033[0;37m"
+    fi
+    if [[ -z "$metric_types" ]]; then
+        is_error=true
+        echo -e "\033[0;31mrun.sh (1): Region for Google Cloud Platform is missing please rerun the script with the relevant parameters\033[0;37m"
```
Comment (Doron-Bargo) [inline run.sh]: Not sure this error is correct

File: run.sh [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+function is_gcloud_install(){
+    echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")]running command gcloud -v .."
+
+    gcloud_ping=`gcloud -v 2>/dev/null | wc -w`
+
+    if [ $gcloud_ping -gt 0 ]
+    then
+        return
+    else
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to download telegraf  ..."
```
Comment (Doron-Bargo) [inline run.sh]: What is the connection to telegraf?

File: run.sh [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+function_name="metrics_gcp"
+
+# Prints usage
+# Output:
+#   Help usage
+function show_help () {
+    echo -e "Usage: ./run.sh --listener_url=<listener_url> --token=<token> --region=<region> --function_name=<function_name> --metric_types=<metric_types>"
+    echo -e " --listener_url=<listener_url>       Logz.io Listener URL (You can check it here https://docs.logz.io/user-guide/accounts/account-region.html)"
+    echo -e " --token=<token>                     Logz.io token of the account you want to ship to."
+    echo -e " --region=<region>                   Region where you want to upload Cloud Funtion."
```
Comment (Doron-Bargo) [inline run.sh]: Not clear which region is this one please change to gcp_region

File: README.md [hunk 5]
```diff
@@ -1 +1,73 @@
...
+./run.sh --listener_url=<listener_url> --token=<token> --region=<region> --function_name=<function_name> --metric_types=<metric_types>
+```
+
+Replace the variables as per the table below:
+
+| Parameter    | Description                                                                                                                                                   |
+| ------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| listener_url | Use the listener URL specific to the region of your Logz.io account. You can look it up [here](https://docs.logz.io/user-guide/accounts/account-region.html). |
+| token        | The metrics' shipping token of the account you want to ship to.                                                                                               |
+| region       | Region where you want to upload Cloud Function. \*`Requires for Deploy to Cloud option for platform`.                                                         |
```
Comment (Doron-Bargo) [inline README.md]: Its need to be more clear its gcp regions and not logzio
Also is it the name of the region or the id?

File: run.sh [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+    echo -e " --help                              Show usage"
+}
+
+# Gets arguments
+# Input:
+#   Client's arguments ($@)
+# Output:
+#   listener_url - Logz.io Listener URL
+#   token - Logz.io token of the account user want to ship to.
+#   metric_type - Region where user want to upload Cloud Funtion.
```
Comment (Doron-Bargo) [inline run.sh]: This comment is not correct

File: run.sh [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+#   type -  Log type. Help classify logs into different classifications
```
Comment (Doron-Bargo) [inline run.sh]: log type?

**Review by Doron-Bargo — CHANGES_REQUESTED** [review]
Great, Please see my comments

File: README.md [hunk 8]
```diff
@@ -1 +1,72 @@
...
+
+<details>1. Log in to your GCP account:
+
+```shell
+gcloud auth login
+```
+
+2. Navigate to the relevant project.
+
+3. Set the `project id` for the project that you want to send logs from:
```
Comment (Doron-Bargo) [inline README.md]: collect metrics from

File: run.sh [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+function is_gcloud_install(){
+    echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")]running command gcloud -v .."
+
+    gcloud_ping=`gcloud -v 2>/dev/null | wc -w`
+
+    if [ $gcloud_ping -gt 0 ]
+    then
+        return
+    else
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to get gcloud CLI. Please install Gcloud and login to proper account from where you want to send metrics..."
```
Comment (Doron-Bargo) [inline run.sh]: Failed to use

File: run.sh [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+        is_error=true
+        echo -e "\033[0;31mrun.sh (1): Metric type is missing please rerun the script with the relevant parameters\033[0;37m"
+    fi
+  
+    if $is_error; then
+        echo -e "\033[0;31mrun.sh (1): try './run.sh --help' for more information\033[0;37m"
+        exit 1
+    fi
+
+    echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")] Validation of arguments passed."
```
Comment (Doron-Bargo) [inline run.sh]: passed successfully

File: run.sh [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+}
+
+# Download/UNZIP/Move Telegraf version as 1.24.4
+# Error:
+#   Exit Code 1
+function download_telegraf(){
```
Comment (Doron-Bargo) [inline run.sh]: Please replace telegraf to Telegraf

File: run.sh [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+    tar -zxf ./$telegraf_zip_name --directory .
+    if [[ $? -ne 0 ]]; then
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to unzip telegraf..."
+        exit 1
+    fi
+    
+    # Move telegraf file to function folder
+    mv $default_telegraf_folder $function_folder
+    if [[ $? -ne 0 ]]; then
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to move telegraf to function folder..."
```
Comment (Doron-Bargo) [inline run.sh]: to the function's folder

File: run.sh [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+    chmod +x $function_folder/telegraf
+    if [[ $? -ne 0 ]]; then
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to give permission to telegraf..."
+        exit 1
+    fi
+
+    # Remove telegraf zip tar.gz file
+    rm -rf ./telegraf-${telegraf_version}
+    if [[ $? -ne 0 ]]; then
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to remove telegraf folder  ..."
```
Comment (Doron-Bargo) [inline run.sh]: remove the Telegraf folder

File: run.sh [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+
+
+# Populate user's credentials to telegraf configuration 
+function populate_data(){
+    echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")] Populating telegraf config ..."
+
+    # Remove telegraf.conf for create clean
+    rm -rf ./function_cloud/telegraf.conf
+    if [[ $? -ne 0 ]]; then
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Error to remove..."
```
Comment (Doron-Bargo) [inline run.sh]: Its not clear

File: run.sh [hunk 15]
```diff
@@ -0,0 +1,447 @@
...
+function get_project_id(){
+    echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")] Getting Google Project ID..."
+
+    gcloud config get-value project
+    if [[ $? -ne 0 ]]; then
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to get user project id  ..."
+        exit 1
+    else
+        project_id="$(gcloud config get-value project)"
+        echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")] Project ID=$project_id where will launch integration."
```
Comment (Doron-Bargo) [inline run.sh]: where integration will be launched

File: run.sh [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+
+    gcloud iam service-accounts keys create credentials.json --iam-account ${account_name}@${project_id}.iam.gserviceaccount.com
+    if [[ $? -ne 0 ]]; then
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to add permissions to service account."
+        exit 1
+    fi
+    # Move telegraf file to function folder
+    mv ./credentials.json $function_folder
+    if [[ $? -ne 0 ]]; then
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to move telegraf to function folder..."
```
Comment (Doron-Bargo) [inline run.sh]: to the function's folder

**Review by Doron-Bargo — CHANGES_REQUESTED** [review]
Please check my comments

File: run.sh [hunk 17]
```diff
@@ -0,0 +1,447 @@
...
+	echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")] Parsed metrics types."
+}
+
+
+# Populate user's credentials to Telegraf configuration 
+function populate_data(){
+    echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")] Populating Telegraf config ..."
+
+    # Remove telegraf.conf for create clean
+    rm -rf ./function_cloud/telegraf.conf
```
Comment (Doron-Bargo) [inline run.sh]: why you need to remove the Telegraf file and not just overwrite it?

**Review by Doron-Bargo — CHANGES_REQUESTED** [review]
One small comment and we are done

File: run.sh (hunk 17 above)
Comment (resdenia) [reply run.sh]: Yes, it's a good improvement.

**Review by Doron-Bargo — APPROVED** [review]
lgtm

**Pull request merged**